R/quosure-join.R

```diff
@@ -0,0 +1,103 @@
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `y` has unique code element placed before common element. This means that `y`
+#' in the environment of the `y` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases to be done
+#' @param x (`Quosure`)
+#' @param y (`Quosure`)
+#' @examples
+#' q1 <- new_quosure(
+#'   code = c(iris1 = "iris1 <- iris", mtcars1 = "mtcars1 <- mtcars"),
+#'   env = list2env(list(
+#'     iris1 = iris,
+#'     mtcars1 = mtcars
+#'   ))
+#' )
+#' q2 <- q1
+#' q1 <- eval_code(q1, "iris2 <- iris")
+#' q2 <- eval_code(q2, "mtcars2 <- mtcars")
+#' qq <- join(q1, q2)
+#' get_code(qq)
+#' @export
+setGeneric("join", function(x, y) {
+  standardGeneric("join")
+})
+
+#' @rdname join
+#' @export
+setMethod("join", signature = c("Quosure", "Quosure"), function(x, y) {
+  join_validation <- check_joinable(x, y)
+
+  # join expressions
+  if (!isTRUE(join_validation)) {
+    stop(join_validation)
+  }
+
+  id_unique <- !y@id %in% x@id
+  x@id <- c(x@id, y@id[id_unique])
+  x@code <- .keep_code_name_unique(x@code, y@code[id_unique])
+
+  # insert (and overwrite) objects from env2 to env
+  x@env <- .copy_env(x@env)
```

Comment:
why do we need to copy env from self into self?

Reply:
imagine you have the situation like this:
```r
x <- new_quosure(env = list2env(a = 1), code = "a <- 1")
y <- new_quosure(env = list2env(a = 2), code = "a <- 2") 

z <- join(x, y)
```

If we didn't copy the environment we join would result with `z` and `x` sharing the same environment (same as in eval_code).

Reply:
Yes you are right. It seems that better naming / docs / ??? is needed.
At the very beginning I was thinking that this is an instance method performing join into itself, i.e. q1.join(q2)
This is however yet another instance constructor basing on two instances, i.e. q_new <- join(q1, q2).
If this should be a constructor - can we introduce `x_new <- new_quosure(...); ...` logic somehow to make it more clear that the new instance is being created.

Reply:
marking as resolved:
I guess this should be clear now Marking as resolved. Please comment the PR if something require improvement 

https://github.com/insightsengineering/teal.code/pull/68/files#diff-3f9293e71b2b2e30dc82142853ab25338b5b3f4a594fe1f326c932a64083aaba